runtime/drivers/salesforce/authentication.go

```diff
@@ -0,0 +1,109 @@
+package salesforce
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+
+	force "github.com/ForceCLI/force/lib"
+)
+
+const defaultEndpoint = "https://login.salesforce.com"
+
+type authenticationOptions struct {
+	Endpoint     string
+	Username     string
+	Password     string
+	JWT          string
+	ConnectedApp string
+}
+
+// Authenticable represents the authentation subset of the force api
+type Authenticable interface {
+	JwtAssertionForEndpoint(endpoint string, username string, keyfile string, clientID string) (string, error)
+	JwtLoginAtEndpoint(endpoint string, assertion string) (*force.Force, error)
+	ForceSoapLoginAtEndpoint(endpoint string, userName string, password string) (*force.Force, error)
+}
+
+// a concrete implementation of Authenticable
+type forceProvider struct{}
+
+func authenticate(options authenticationOptions) (*force.Force, error) {
+	if options.ConnectedApp == "" {
+		options.ConnectedApp = defaultClientID
+	}
+	force.ClientId = options.ConnectedApp
+	provider := forceProvider{}
+
+	isUsernameSelected := len(options.Username) > 0
+	isJWTSelected := len(options.JWT) > 0
+	isSOAPSelected := len(options.Password) > 0
+
+	endpoint, err := endpoint(options)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case isJWTSelected:
+		return jwtLogin(endpoint, options, provider)
+	case isSOAPSelected:
+		return provider.ForceSoapLoginAtEndpoint(endpoint, options.Username, options.Password)
+	case !isUsernameSelected:
+		return nil, fmt.Errorf("username missing")
```

Comment:
Can this error be thrown earlier ? I see `Username` is being used in both `jwtLogin` and `ForceSoapLoginAtEndpoint`.

Reply:
Addressed in 8b973d169b894a27a2792ebf344c828ad452941a.